File: ck_pojie.py
"""
cron: 53 11 * * *
new Env('吾爱破解');
"""

from utils import get_data
import re, time, random, shutil, tempfile
from notify_mtr import send
from datetime import datetime
from selenium import webdriver
from fake_useragent import UserAgent
from selenium_stealth import stealth
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import NoSuchElementException, TimeoutException, WebDriverException

def setup_browser():
    user_data_dir = tempfile.mkdtemp(prefix="selenium_chrome_")
    options = webdriver.ChromeOptions()
    options.add_argument("--headless")
    options.add_argument("--no-sandbox")
    options.add_argument("--disable-dev-shm-usage")
    options.add_argument(f"--user-data-dir={user_data_dir}")
    options.add_argument(f"--user-agent={UserAgent().chrome}")
    options.add_argument("--disable-gpu")  # 新增：解决 GPU 初始化错误
    options.add_argument("--disable-software-rasterizer")
    options.add_argument("--disable-blink-features=AutomationControlled")
    options.add_experimental_option("excludeSwitches", ["enable-automation"])

    service = webdriver.ChromeService()
    service.executable_path='/usr/bin/chromedriver'
    service.service_args = [
        '--verbose',                    # 详细日志
        '--readable-timestamp',         # 可读时间
        '--enable-chrome-logs',         # 启用 Chrome 内部日志
        '--no-sandbox',                 # 容器中常用
        '--disable-dev-shm-usage',      # 防止内存溢出
        # '--append-log',               # 如需追加日志
        '--log-path=/tmp/52pojie.log'
    ]

    driver = webdriver.Chrome(options=options, service=service)
    stealth(driver,
        platform="Win32",
        fix_hairline=True,
        hide_webdriver=True,
        vendor="Google Inc.",
        webgl_vendor="Intel Inc.",
        languages=["zh-CN", "zh"],
        renderer="Intel Iris OpenGL Engine",
    )
    actions = ActionChains(driver)
    actions.move_by_offset(random.randint(10, 100), random.randint(10, 100)).perform()

    # 禁用 debugger 语句
    driver.execute_script("""
        (function() {
            window.__proto__.__defineGetter__('debugger', () => {});
            Object.defineProperty(window, 'debugger', {value: undefined, writable: false});
            let originalSetInterval = window.setInterval;
            window.setInterval = function(callback, timeout) {
                if (typeof callback === 'function' && callback.toString().includes('debugger')) return;
                if (typeof callback === 'string' && callback.includes('debugger')) return;
                return originalSetInterval.apply(this, arguments);
            };
            let originalSetTimeout = window.setTimeout;
            window.setTimeout = function(callback, timeout) {
                if (typeof callback === 'function' && callback.toString().includes('debugger')) return;
                if (typeof callback === 'string' && callback.includes('debugger')) return;
                return originalSetTimeout.apply(this, arguments);
            };
        })();
    """)

    return driver, user_data_dir

def sign(cookie, i):
    msg = ''
    driver, user_data_dir = setup_browser()

    try:
        driver.get('https://www.52pojie.cn/forum.php')
        for single_cookie in cookie.split('; '):
            name, value = single_cookie.split('=', 1)
            driver.add_cookie({'name': name, 'value': value})
        time.sleep(2)

        driver.get('https://www.52pojie.cn/forum.php')
        # driver.get('https://visit.zjsru.edu.cn/visitor/qrCode?id=fec1f2b23ce1f0bb1a595423af169b79')


        if '自动登录' in driver.page_source:
            return f'❌ 无法登录！可能Cookie失效，请重新修改'

        name_element = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, 'strong.vwmy a[href*="uid="]'))
        )
        um_element = driver.find_element(By.CSS_SELECTOR, '#um p:nth-of-type(2)')
        msg = f"---- {name_element.text} 吾爱破解 签到状态 ----\n"
        upmine = driver.find_element(By.ID, "g_upmine").get_attribute("textContent").strip()
        integral = driver.find_element(By.ID, "extcreditmenu").get_attribute("textContent").strip()

        if len(driver.find_elements(By.CSS_SELECTOR, 'img.qq_bind[src*="wbs.png"]')) > 0:
            return f"{msg}<b><span style='color: green'>✅ 今天已经签到过了</span></b>\n{integral} | {upmine}"

        try:
            sign_button = driver.find_element(By.CSS_SELECTOR, 'img.qq_bind[src*="qds.png"]')
            driver.execute_script("arguments[0].scrollIntoView();arguments[0].click()", sign_button)
            time.sleep(8)
            if len(driver.find_elements(By.CSS_SELECTOR, 'img.qq_bind[src*="wbs.png"]')) > 0:
                return f"{msg}✅ 签到成功\n{integral} | {upmine}"
        except NoSuchElementException:
            return f"{msg}\n❌ 签到失败"

    except TimeoutException as e:
        return f"{msg}<b><span style='color: red'>超时异常：</span></b>\n{e}"
    except NoSuchElementException as e:
        return f"{msg}<b><span style='color: red'>签到失败：</span></b>\n{e}"
    except WebDriverException as e:
        return f"{msg}<b><span style='color: red'>WebDriver异常：</span></b>\n{e}"
    except Exception as e:
        return f"{msg}<b><span style='color: red'>未知异常：</span></b>\n{e}"
    finally:
        try:
            with open(f"/tmp/52pojie_{i}.html", "w", encoding="utf-8") as f:
                f.write(driver.page_source)
            driver.save_screenshot(f"/tmp/52pojie_{i}.png")
        except:
            pass
        driver.quit()
        shutil.rmtree(user_data_dir, ignore_errors=True)

def main():
    _data = get_data()
    check_items = _data.get("POJIE", [])

    msg_all = ""
    for i, check_item in enumerate(check_items, start=1):
        cookie = check_item.get("cookie")
        sign_msg = sign(cookie, i)
        msg = f"账号 {i} 签到状态: {sign_msg}"
        msg_all += msg + "\n\n"

    return msg_all

if __name__ == "__main__":
    result = main()
    # send("吾爱破解", result)
    print(result)

File: ck_pojie.js
/*
吾爱破解签到
cron "2 0,11 * * *" ck_pojie.js
*/

const fs = require('fs');
const notify = require('./notify')
const { chromium } = require('playwright');

const { Env, networkLog, MagicJS, getData } = require('./utils');
const $ = new Env('吾爱破解 签到');
const magicJS = MagicJS('吾爱破解', 'INFO');
const COOKIES_POJIE = getData().POJIE;

async function setupBrowser() {
    const browser = await chromium.launch({
        headless: true,
        ignoreHTTPSErrors: true,
        executablePath: '/usr/bin/chromium-browser',
        args: [
            '--no-sandbox',
            '--disable-gpu',
            '--ignore-gpu-blocklist',
            '--disable-dev-shm-usage',
            '--disable-software-rasterizer',
            '--disable-blink-features=AutomationControlled'
        ]
    });

    const initScript = `
        Object.defineProperty(navigator, 'webdriver', { get: () => false });
        Object.defineProperty(navigator, 'platform', { get: () => 'Win32' });
        delete window.debugger;
        const filter = f => !(typeof f === 'function' && f.toString().includes('debugger'));
        const st = window.setTimeout, si = window.setInterval;
        window.setTimeout = (cb, t) => filter(cb) && st(cb, t);
        window.setInterval = (cb, t) => filter(cb) && si(cb, t);
    `;
    const context = await browser.newContext({
        baseURL: 'https://www.52pojie.cn',
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        extraHTTPHeaders: {
            'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        },
    });
    await context.addInitScript(initScript);

    return { browser, context };
}

async function sign(context, cookie, index) {
    const signKey = `pojie_sign_${index}`;
    if (magicJS.read(signKey) === magicJS.today()) return `账号 ${index}: ✅ 今日已签到`;

    let msg = '';
    const msgHead = `账号 ${index}: `;
    const page = await context.newPage();
    const opts = { waitUntil: 'networkidle', timeout: 20000 };
    const cookies = cookie.split(/;\s*/).map(c => {
        const [name, ...rest] = c.split('=');
        return { name, value: rest.join('='), domain: '.52pojie.cn', path: '/', sameSite: 'Lax' };
    });
    await context.clearCookies();
    await context.addCookies(cookies);

    const logger = await networkLog(page, {
        filter: ['forum.php', 'task', 'apply'],
        filename: `52pojie_log_${index}.json`,
    });

    try {
        await page.goto('/forum.php', opts);
        const [upmine, integral, username] = await Promise.allSettled([
            page.locator('#g_upmine').textContent({ timeout: 1500 }),
            page.locator('#extcreditmenu').textContent({ timeout: 1500 }),
            page.locator('strong.vwmy a[href*="uid="]').textContent({ timeout: 1500 })
        ]).then(r => r.map(x => x.status === 'fulfilled' ? x.value : null));
        if (!username) return `${msgHead} ❌ Cookie 失效`;

        msg += `${msgHead}${username}\n`;
        const qds = page.locator('#um img[src*="qds.png"]');
        if (await qds.isVisible({ timeout: 2000 })) {
            const href = await qds.first().evaluate(img => img.closest('a')?.href);
            await page.goto(href, opts);
            await page.waitForSelector('#um img[src*="wbs.png"]', { timeout: 5000 })
                .catch(() => null);

            await page.goto('/forum.php', opts);
            const signed = await page.locator('#um img[src*="wbs.png"]');
            if (signed.isVisible({ timeout: 2000 })) {
                msg += `${magicJS.today()} ✅ 签到成功\n`;
                magicJS.write(signKey, magicJS.today());
            }
        } else {
            msg += `${magicJS.today()} ℹ️ 已签到过\n`;
        }

        await page.goto('/home.php?mod=spacecp&ac=credit&op=base', opts);
        const wuaibi = (await page.locator('li:has(em:has-text("吾爱币"))').innerText({ timeout: 2000 })
            .catch(() => '')).match(/\d+\s*CB/)?.[0] || '未知';
        msg += `积分: ${integral} | 吾爱币: ${wuaibi} | 威望: ${upmine}\n`;

        fs.writeFileSync(`/tmp/52pojie_${index}.html`, await page.content());
        await page.screenshot({ path: `/tmp/52pojie_${index}.png`, fullPage: true });
        logger.stop();
    } catch (err) {
        msg += `❌ 异常: ${err.message}\n`;
    } finally {
        await page.close().catch(() => {});
    }

    return msg;
}

async function main() {
    const { browser, context } = await setupBrowser();

    let msgAll = '=== 吾爱破解 签到结果 ===\n';

    for (let i = 0; i < COOKIES_POJIE.length; i++) {
        const cookie = COOKIES_POJIE[i]?.cookie;
        let res = cookie ? await sign(context, cookie, i + 1)
            : `账号 ${i + 1}: ❌ Cookie 为空`;

        msgAll += res;
        if (i < COOKIES_POJIE.length - 1) await magicJS.sleep(2000);
    }

    $.log(msgAll);
    magicJS.done();
    if (/成功|异常|失效/.test(msgAll)) notify.sendNotify('吾爱破解 签到', msgAll);

    await context.close().catch(() => {});
    await browser.close().catch(() => {});
}

main().catch(err => {
    $.logErr('❌ 脚本异常:', err);
    process.exit(1);
});

module.exports = { main };
